snapcraft/commands/store/_legacy_account.py

```diff
@@ -0,0 +1,164 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Backwards compatibility for login --with."""
+
+import base64
+import configparser
+import os
+from pathlib import Path
+from typing import Optional, Sequence
+
+import craft_store
+import pymacaroons
+import xdg.BaseDirectory
+from overrides import overrides
+from urllib3.util import parse_url
+
+from snapcraft import errors
+
+from . import constants
+
+
+def _load_potentially_base64_config(config_content: str) -> configparser.ConfigParser:
+    parser = configparser.ConfigParser()
+    try:
+        parser.read_string(config_content)
+    except configparser.Error as parser_error:
+        # The config may be base64-encoded, try decoding it
+        try:
+            decoded_config_content = base64.b64decode(config_content).decode()
+        except base64.binascii.Error as b64_error:  # type: ignore
+            # It wasn't base64, so use the original error
+            raise errors.SnapcraftError(
+                f"Cannot parse config: {parser_error}"
+            ) from b64_error
+
+        try:
+            parser.read_string(decoded_config_content)
+        except configparser.Error as new_parser_error:
+            raise errors.SnapcraftError(
+                f"Cannot parse config: {parser_error}"
+            ) from new_parser_error
+
+    return parser
+
+
+def _deserialize_macaroon(value) -> pymacaroons.Macaroon:
+    try:
+        return pymacaroons.Macaroon.deserialize(value)
+    except:  # noqa LP: #1733004
+        raise errors.SnapcraftError(  # pylint: disable=raise-missing-from
+            "Failed to deserialize macaroon"
+        )
+
+
+def _macaroon_auth(conf) -> str:
+    """Format a macaroon and its associated discharge.
+
+    :return: A string suitable to use in an Authorization header.
+    """
+    host = parse_url(constants.UBUNTU_ONE_SSO_URL).host
+    root_macaroon_raw = conf.get(host, "macaroon")
+    if root_macaroon_raw is None:
+        raise errors.SnapcraftError("Root macaroon not in the config file")
+    unbound_raw = conf.get(host, "unbound_discharge")
+    if unbound_raw is None:
+        raise errors.SnapcraftError("Unbound discharge not in the config file")
+
+    root_macaroon = _deserialize_macaroon(root_macaroon_raw)
+    unbound = _deserialize_macaroon(unbound_raw)
+    bound = root_macaroon.prepare_for_request(unbound)
+    discharge_macaroon_raw = bound.serialize()
+    auth = f"Macaroon root={root_macaroon_raw}, discharge={discharge_macaroon_raw}"
+
+    return base64.b64encode(auth.encode()).decode()
+
+
+def get_auth(config_content: str) -> str:
+    """Return a legacy authorization header."""
+    conf = _load_potentially_base64_config(config_content)
+    auth = _macaroon_auth(conf)
+
+    return auth
+
+
+class LegacyUbuntuOne(craft_store.UbuntuOneStoreClient):
+    """Legacy client to easily transition existing CI users."""
+
+    _CONFIG_PATH = Path(xdg.BaseDirectory.xdg_config_home) / "snapcraft/ci.cfg"
```

Comment:
Maybe use `legacy_auth.cfg` or something that's not CI-specific?

Reply:
updated